Add support for exclusion rules in dockerignore

buddhamagnet opened the pull request:
Parallel pull request to https://github.com/docker/docker/pull/11823 (since closed).

buddhamagnet committed:
add support for exclusion rules in dockerignore

Signed-off-by: Dave Goodchild <buddhamagnet@gmail.com>

GordonTheTurtle commented:
Can you please sign your commits following these rules:

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

~~~console
$ git clone -b "dockerignore-ignores" git@github.com:buddhamagnet/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

buddhamagnet commented:
Amended last commit, now signed.

icecrime commented:
@buddhamagnet Thanks! There's a formatting problem to one of your files:

```
These files are not properly gofmt'd:
 - pkg/fileutils/fileutils_test.go

Please reformat the above files using "gofmt -s -w" and commit the result.
```

Can you please fix it?

buddhamagnet commented:
Done, thanks.

cpuguy83 commented:
Can you squash and make sure the commit is signed? Tests aren't running because your last commit isn't signed. These should be 1 commit anyway.
Lmk know if you need any help with this!

Thanks!

buddhamagnet commented:
Ah sorry I committed at work and hadn't updated my local git config. On it.

buddhamagnet commented:
Done that, apologies. Don't have time to do the squash now will do later, but the commits are now both signed.

thaJeztah commented:
I see the tests failed

buddhamagnet commented:
My bad, had added a failing test for the single exclamation point case. Will complete that locally, apologies.

buddhamagnet commented:
Now rebased against master and squashed.

thaJeztah commented:
Sorry, @buddhamagnet looks like it still failed; https://jenkins.dockerproject.com/job/Docker-PRs/5544/console

(The Windows fail doesn't look like your fault)

duglin commented on pkg/fileutils/fileutils.go:
```diff
@@ -7,21 +7,50 @@ import (
+				continue
+			}
+
 			if filepath.Clean(relFilePath) == "." {
```
Should this check for "." apply to the ! case too?
If so, you can move this up before the "if negative" line and then the line below becomes:
```
matched  = !negative
```
and you can remove the "if negative" check.

buddhamagnet replied:
Yep - done.

duglin commented on pkg/fileutils/fileutils.go:
```diff
@@ -7,21 +7,50 @@ import (
+
+		if pattern[0] == '!' {
+			if len(pattern) == 1 {
+				continue
```
Where did we end up on this discussion?  I was thinking we should generate an error instead of silently ignore these lines.

duglin replied:
Let's also add a test for this as well as the "." case below to make sure someone doesn't remove these checks by mistake.

buddhamagnet replied:
@duglin I am working on it. Have the test stacked up but don't want to break the build again until I'm sure.

duglin commented on pkg/fileutils/fileutils_test.go:
```diff
@@ -0,0 +1,26 @@
+	if match != false {
+		t.Errorf("failed to get false match on exclusion pattern, got %v", match)
+	}
+}
```
Can we add:
```
... { "!fileutils.go", "*go" }
```
I believe the result should be "true". Right?

buddhamagnet replied:
Test added.

duglin commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -652,17 +654,20 @@ is not recursive.
 Currently there is no support for regular expressions. Formats
-like `[^temp*]` are ignored.
+like `[^temp*]` are ignored. The second to last line adds a rule to ignore all
+markdown files and the subsequent line adds an exlusion rule for LICENCE.md.
```
Let's add something about how order impacts it - e.g.:
```
It is important to point out that the order in which the "!" lines appear will influence the matching algorithm; the last line of the .dockerignore that matches will determine whether a file is included or not. For example, in the above example, if the last two lines were reversed then LICENSE.md would be excluded from the build because the *.md would add all "md" files back into the ignore list. In other words, the !LICENSE.md line would have no effect at all because of the subsequent *.md.
```
that is correct isn't it?

buddhamagnet replied:
Yes makes sense. Done.

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -3695,6 +3695,49 @@ func TestBuildDockerignore(t *testing.T) {
+	dockerfile := `
+        FROM busybox
+        ADD . /bla
+               RUN [[ -f /bla/src/x.go ]]
```
why are these indented more?

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -3695,6 +3695,49 @@ func TestBuildDockerignore(t *testing.T) {
+		"docs/config.json": "{ config: 1 }",
+		".dockerignore": `.git
+pkg
+gitignore
```
should this be ".gitignore" ?  (the dot)

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -3695,6 +3695,49 @@ func TestBuildDockerignore(t *testing.T) {
+               RUN [[ ! -e /bla/README.md ]]
+               RUN [[ ! -e /bla/.git ]]
+               RUN [[ -f /bla/LICENSE.md ]]
+               RUN [[ -f /bla/docs/config.json]]`
```
let's add another (non-md) file under "docs/" and make sure its not there.  I was going to suggest we test for "useless.md" but that's excluded by the *.md.  I want to make sure the "docs" line catches everything except config.json.

duglin replied:
I think you missed this comment.

buddhamagnet commented:
@thaJeztah apologies I didn't write the integration part. I was on this with @omeid but can't seem to reach him (@omeid hope all is well) so I created a fresh PR from master to move things forward. Will work on all this early next week and hopefully get it into shape.

Out of interest @duglin @cpuguy83 why the insistence on squashing all commits during the review process. My common process is to rebase and squash on approval before merge. Just curious.

thaJeztah commented:
@buddhamagnet no problem; on the squashing: all tests are run for *each individual commit*, so this can result in longer test-times and failing tests if commits are not complete

duglin commented:
@buddhamagnet I don't care much about having one vs many commits during the review cycle, as long as its done before its merged. The thing I worry about is that reviewers may jump the gun and merge before you do your final "clean-up" and squash them down to one so I prefer to have the PR in as "final" a state as possible.

buddhamagnet commented:
@thaJeztah so can I just commit --amend on each iteration and force push? Is that an acceptable workflow?

duglin commented:
@buddhamagnet yup - that's what I do.

thaJeztah commented:
For completeness, there's some information on the squashing in the [conventions section](https://github.com/docker/docker/blob/master/CONTRIBUTING.md#conventions) of CONTRIBUTING.md

buddhamagnet commented:
I've removed that integration test for now as I'm not sure why it's failing and would prefer to cover the feature with unit tests in fileutils_test.go, round out the work and then add the integration test at the end?

duglin commented:
I think the integration test was failing because of the missing dot on .gitignore

duglin commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -652,17 +654,27 @@ is not recursive.
+influence the matching algorithm; the last line of the .dockerignore that matches
+will determine whether a file is included or not. For example, in the above example,
+if the last two lines were reversed then LICENSE.md would be excluded from the build
+because the *.md would add all "md" files back into the ignore list. In other words,
```
add back-ticks ( ` ) around the file snippets on these lines.

buddhamagnet replied:
Done!

buddhamagnet commented:
@duglin +1 on the integration test, added it back in. Apologies, new to this project.

buddhamagnet commented:
Nope it ain't liking that integration test.

moxiegirl commented:
@buddhamagnet  We have a formal PR review process which is explained here:  http://docs.docker.com/project/review-pr/  You can also track where you are in the process by your PR labels (I have to document these they recently changed).  Right now, you are at step 2 which is code review --- the next step is 3 which is docs review.  Hope this helps.

buddhamagnet commented:
Taken it out again and it stays out until I'm ready!

buddhamagnet commented:
So tests passing again will try and keep it that way!

buddhamagnet commented:
Now return an error for the single ! pattern in fileutils.go, backed by a test.

buddhamagnet commented:
OK now we're getting somewhere. The single ! now returns an error, we have a test for it, and the tests are passing. @duglin I now need some advice on where to handle this error - in builder/evaluator.go or somewhere else? Still finding my way around the codebase but we're getting closer.

buddhamagnet commented:
I've now squashed this down into a single commit and will amend and force push from this point on if we are getting close to mergeability.

buddhamagnet commented:
Also converted those two checks for "" and ! into small non-exported functions for readability.

duglin commented:
@buddhamagnet I kind of like the idea of putting the check for "!" into ReadDockerIgnore() then its only checked once and not each time we loop over the excludes list.  The only downside to this is that if someone calls Matches() w/o going thru ReadDockerIgnore() then we'll miss the check.  If Matches() was put into a builder file then I think it would be safer, but since its in a generic fileutils we're almost asking for it to be called from something other than the builder.  For now, let's just leave it.

duglin commented:
What's the status of the integration-cli test?  The rest looks good.

buddhamagnet commented:
@duglin is this good to merge if the integration test passes you think?

duglin commented:
I was kind of hoping you'd add back in your test

buddhamagnet commented:
@duglin that's what I meant - have added the integration test back with the correction.

buddhamagnet commented:
Think I see the issue now, not just the missing dot. Shell command was returning error code 2 on creating docs/config.json think it's just the missing space between the filename and the brackets.

buddhamagnet commented:
Odd the integration test is failing and I can't quite work out why. It's returning an error code 1 creating the config.json file, I corrected the sh syntax error which was throwing a 2. Anyone got any immediate ideas from the patterns in the other tests? Will keep digging this end.

buddhamagnet commented:
Anyway burgers and beers are calling will check in later.

duglin commented:
@buddhamagnet I think what's going on is that when it checks the "docs" dir it, correctly, decides that it needs to be excluded.  So, look at this line:
https://github.com/docker/docker/blob/master/pkg/archive/archive.go#L469
that will tell the Walk() func to skip that dir - which means it never gets a chance to look at docs/config.json to re-include it.

Not sure yet how to fix this. But one option, and we'll need to check, is to remove that SkipDir return so that it does walk into the dir.  But then our Match() func will need to be smarter and do remove dirs (and its children) itself instead of counting on the walk func skipping them.

buddhamagnet commented:
@duglin so is this a problem with the new code that has been surfaced by the integration test or a problem with the integration test itself?

buddhamagnet commented:
OK I'll wait to get some direction on how you'd like that solved.

duglin commented:
Depends on your point of view. The old code worked just fine because it could safely assume that once you excluded a dir all of its children will always be excluded.   This new feature changes that assumption.  I think we may have to do what I mentioned before, not skip entire dirs and then modify our match logic to handle dir manually instead of relying on ‘go’ to do it for us.

Haven’t fully thought this through yet but I was thinking that it would work like this:
- given a file path - say: dir1/file
- we want to see if it should be included or excluded based on a .dockerignore of:
	dir1 
	!dir1/file
- as we loop over the patterns we see that “dir1” tells us to excluded it, so as of now “match” is true.  However, we need to make sure that this check is smart and doesn’t just check the prefix of the file path, it needs to do something like: file == pattern || file.HasSuffix(pattern + “/“)  - checking to make sure pattern doesn’t already have a trailing /.  Ug, and taking into account windows vs linux paths separators. 
- then we hit the 2nd pattern and since “file == pattern”, and we have “negative == true”, “match” will now become false.
so I think the net result will be dir1/file will be included in the build context
But only coding it up will know for sure if its really this simple.

duglin commented:
Of course, making it also check for wildcard will be a joy :-)

thaJeztah commented:
Perhaps it might be worth looking around if there's some existing library that provides this; can't imagine this is the first time someone tries to implement this?

buddhamagnet commented:
Removing that test again and making some small amendments to the Matches function to clean it up a little before stepping back and having a deeper think about the way forward.

buddhamagnet commented:
Rebased against master and force pushed.

duglin commented:
Did you address the issue exposed by the integration-cli testcase?

buddhamagnet commented:
@duglin no did you read the comment?

duglin commented:
yep I read it, but that was a whole day ago  :-)  was hoping for some exciting news

buddhamagnet commented:
@duglin it's open source baby time is limited ;(

buddhamagnet commented:
@duglin I'm new to this project, going to take some time to follow through the implications of the change and understand what's going on in archive.go etc.

duglin commented:
no worries!  thanks for the effort

buddhamagnet commented:
@duglin arghhh now I'm in at the deep end. Have boot2docker running on my Mac, built a container from my branch and mounted against it, built Docker on the Mac, jumped onto the container and stuck the binary in /usr/bin, created a dummy application with a Dockerfile and .dockerignore and built it, so can see the dockerignore code is working, now trying to figure out how to resolve the issue raised in the integration test!

duglin commented:
@buddhamagnet 
Will have more time tonight but for now, I think you need to start by removing the if-statement here: https://github.com/docker/docker/blob/master/pkg/archive/archive.go#L452
What that will do is make it so that instead of skipping all dirs (and children) listed in the .dockerinclude it will just skip the dir itself, but still walk into it and examine each child.

Once we have that, we then need to look at the Matches() on https://github.com/docker/docker/blob/master/pkg/archive/archive.go#L452 and probably modify the logic here to be smarter. By that I mean, right now if the .dockerignore has just `dir` then Matches() call will not flag `dir/file` as a match. This wasn't an issue before because of that skip dir stuff I mentioned above.  What this means is that we now need to manually do this check.

What I was thinking is this:
- walk thru the excludes patterns
- if its a direct match (https://github.com/docker/docker/blob/master/pkg/fileutils/fileutils.go#L16) then we can skip it
- if the current file's prefix matches the patten then we can skip it - *iff* the match works when the patterns ends with a / (meaning add a / before we check).  This will be tricky because of wildcards in the pattern - so we may need count the number of dirs in the pattern and extract that same number of dirs from the file being checked - and then do a Match() check.

But as I said, i'm not 100% sure yet - and only coding it up will expose issues.
Hope this helps at little.

duglin commented:
before you go much further you may want to rebase

duglin commented:
@vbatts before we do too far.... what are your thoughts around modifying the archiver code?  Making it support this will require a change in there - see my previous comment: https://github.com/docker/docker/pull/12245#issuecomment-93559143

duglin commented:
@buddhamagnet 
See what you think of these edits:
https://github.com/buddhamagnet/docker/compare/dockerignore-ignores...duglin:Dockerignore
it seems to work but we'd still need to get @vbatts thoughts.

buddhamagnet commented:
Just rebased and force pushed as I see we have new code in the fileutils test.

buddhamagnet commented:
@duglin rebased and slotted those changes in and stepped through and it looks good. Once tests are green I'll remove the SkipDir line entirely.

buddhamagnet commented:
@duglin all the tests pass but I'm having some issue with these lines:

```txt
daemon/daemon.go:824: undefined: fileutils.ReadSymlinkedDirectory
daemon/daemon.go:835: undefined: fileutils.ReadSymlinkedDirectory
daemon/daemon.go:953: undefined: fileutils.CopyFile
daemon/info.go:65: undefined: fileutils.GetTotalUsedFds
```

I rebased against master and force pushed and the new functions in fileutils are all present and correct locally and correctly referenced by daemon.go. Any ideas?

buddhamagnet commented:
@duglin ah must have been some glitchery, it's green now. The windows test is red, but that seems to be fairly normal?

buddhamagnet commented:
@duglin at work now, but can also rebuild the binary later at home and build a dummy app with a Dockerfile and .dockerignore I have set up to test the end result.

buddhamagnet commented:
Removed commented out SkipDir in archive.go.

buddhamagnet commented:
@duglin OK stepped through this manually in my local setup. Built the binary, moved it into the appropriate location on my container, stepped into the dummy app with Dockerfile and .dockerignore, ran docker build, fired up that container and hey presto the fairly sophisticate set of ignore rules is respected.

duglin commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -893,7 +904,6 @@ Creates a new container.
       --cgroup-parent=""         Optional parent cgroup for the container
       --cidfile=""               Write the container ID to the file
       --cpuset-cpus=""           CPUs in which to allow execution (0-3, 0,1)
-      --cpuset-mems=""           Memory nodes (MEMs) in which to allow execution (0-3, 0,1)
```
Something weird is going on with your rebase - this edit isn't part of your stuff.
Same for the stuff down below.

buddhamagnet replied:
No idea where that came from - I pulled master, did a rebase, resolved the conflict in fileutils.go, and force pushed.

duglin replied:
Not sure either since when I tried to rebase it too, just now, those diffs still show up.  I'm not enough of a git expert to know what happened, or how to fix it, aside from extracting the latest versions of those files and applying your changes again.  Or I guess just copy-n-paste those lines that show up in the diff from master.

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -3675,15 +3675,35 @@ func TestBuildDockerignore(t *testing.T) {
 		RUN [[ ! -e /bla/src/_vendor ]]
 		RUN [[ ! -e /bla/.gitignore ]]
 		RUN [[ ! -e /bla/README.md ]]
+		RUN [[ ! -e /bla/dir/dir/f1 ]]
```
Let's add your integration test that you removed.  I really think we need to test the heck out of this. In particular, that new match() logic with wildcards and multi-level dir patterns, etc...   Either here or in the unit tests.

duglin replied:
Right now the Windows tests are failing here.  I would recommend that you add something like:
`RUN find /bla`
after the "ADD" to show all of the files in the /bla dir so its easier to debug when things go wrong.

buddhamagnet replied:
Added integration test back in.

duglin commented on pkg/fileutils/fileutils.go:
```diff
@@ -1,33 +1,77 @@
 		}
-		if matched {
+		if match {
 			if filepath.Clean(relFilePath) == "." {
```
I didn't get a chance to try this, but see if you move this if-statement to the start of this func because technically we should only need to check for this once, and not within the loop.  Hopefully, it doesn't break stuff.  Again, lots of testing needed on this case too because I'm pretty sure we pass "." in the "includes" list, which is what this will deal with.

duglin commented on pkg/fileutils/fileutils.go:
```diff
@@ -1,33 +1,77 @@
+			}
+		} else if !match {
+			// Check to see if the pattern matches one of our parent dirs
+			path := filepath.Dir(relFilePath)
```
these two lines can probably move out of the loop so that they're only done once. Sorry, I should have done more clean-up last night but I was low on time and just wanted to see if the general idea was even possible.

duglin commented on pkg/fileutils/fileutils.go:
```diff
@@ -1,33 +1,77 @@
+			path := filepath.Dir(relFilePath)
+			pathDirs := strings.Split(path, "/")
+			patternDirs := strings.Split(pattern, "/")
+			if path != "." && len(patternDirs) <= len(pathDirs) {
```
if those 2 lines above are moved, then the `path != "."` can probably be moved to the `} else if !match {` line so we don't bother to even enter this chunk of code at all.

buddhamagnet commented:
Re-added the initial integration test I removed, let's see that pass and then we can fine tune and add masses of unit tests.

buddhamagnet commented:
@duglin OK integration tests passed with that one added back in.

buddhamagnet commented:
@duglin not sure how that cruft added up in here after the rebase, seemed clean to me. Amending that then we have green tests and a clean diff to start adding more tests.

buddhamagnet commented:
@duglin OK that integration test failed and I'm out of time again ;(

buddhamagnet commented:
@duglin out of time again. I've removed my older integration test again because you added tests for exlcusions in the test above did you not?

buddhamagnet commented:
@duglin made the initial rework of the logic to extract the check for . from the pattern loop.

buddhamagnet commented:
OK @duglin tests pass with that change. Now cleaning up the inline documentation a little.

buddhamagnet commented:
Added more unit tests.

buddhamagnet commented:
OK moving that conditional check out of the loop broke the integration tests so moved it back, split the integration test into one with a flat dockerignore file and one with exception patterns for clarity. All new unit tests and the split integration tests are passing.

duglin commented:
@buddhamagnet before you go any further I would really fix it so those mystery edits are gone - it won't be merged with them in there.

buddhamagnet commented:
@duglin I don't see the mystery edits in the diff any more they are still an issue?

duglin commented:
hmm you don't see this?
https://github.com/docker/docker/pull/12245/files#diff-c13a5c583fae16a859a034cdd06c7c58L896
https://github.com/docker/docker/pull/12245/files#diff-c13a5c583fae16a859a034cdd06c7c58L1213
https://github.com/docker/docker/pull/12245/files#diff-c13a5c583fae16a859a034cdd06c7c58L1849

buddhamagnet commented:
@duglin my bad can see them no idea how they ended up in here my git rebase flow is pretty solid.

buddhamagnet commented:
@duglin have completed the first pass cleanup, may need a little support if there is anything else that needs cleaning as I want to be careful here. It's weird that the weird edits contain references to cpu-mems etc I didn't see anything odd in the rebase, just some manual conflicts I needed to resolve from upstream changes to fileutils.go during the rebase session.

duglin commented:
Wasn't sure if "first pass" meant you were working on a 2nd pass, but just in case you're not.... there are other weird edits:
https://github.com/docker/docker/pull/12245/files#diff-c13a5c583fae16a859a034cdd06c7c58L1213
https://github.com/docker/docker/pull/12245/files#diff-64d1c3a97fd8b1abf170680883e81aaaL4658
https://github.com/docker/docker/pull/12245/files#diff-64d1c3a97fd8b1abf170680883e81aaaL5558
https://github.com/docker/docker/pull/12245/files#diff-64d1c3a97fd8b1abf170680883e81aaaL5576
https://github.com/docker/docker/pull/12245/files#diff-64d1c3a97fd8b1abf170680883e81aaaL5609

buddhamagnet commented:
@duglin yep I'm aiming to get all of this cleaned up today.

buddhamagnet commented:
@duglin think that's it, put that down to a weird git wormhole.

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -4655,7 +4720,7 @@ func TestBuildExoticShellInterpolation(t *testing.T) {
 
 	_, err := buildImage(name, `
 		FROM busybox
-		
```
can you fix this one - even though its meaningless.  Just to reduce the noise.

buddhamagnet replied:
done sir.

buddhamagnet commented:
@duglin last amendment made, all unit and integration tests green. Windows build failed but it seems to do that most of the time.

duglin commented:
I still think you can move stuff out of the loops. I just tried this code below and all tests passed:
```
func Matches(relFilePath string, patterns []string) (bool, error) {
	matched := false

	if filepath.Clean(relFilePath) == "." {
		// don't let them exclude everything, kind of silly
		return false, nil
	}

	parentPath := filepath.Dir(relFilePath)
	parentPathDirs := strings.Split(parentPath, "/")

	for _, pattern := range patterns {
		negative := false
		if empty(pattern) {
			continue
		}

		pattern = filepath.Clean(pattern)
		if exclusion(pattern) {
			if len(pattern) == 1 {
				logrus.Errorf("Illegal exclusion pattern: %s", pattern)
				return false, errors.New("Illegal exclusion pattern: !")
			}
			negative = true
			pattern = pattern[1:]
		}

		match, err := filepath.Match(pattern, relFilePath)
		if err != nil {
			logrus.Errorf("Error matching: %s (pattern: %s)", relFilePath, pattern)
			return false, err
		}

		if !match && parentPath != "." {
			// Check to see if the pattern matches one of our parent dirs
			patternDirs := strings.Split(pattern, "/")
			if len(patternDirs) <= len(parentPathDirs) {
				match, _ = filepath.Match(strings.Join(patternDirs, "/"),
					strings.Join(parentPathDirs[:len(patternDirs)], "/"))
			}
		}

		if match {
			matched = !negative
		}
	}

	if matched {
		logrus.Debugf("Skipping excluded path: %s", relFilePath)
	}
	return matched, nil
}
```

duglin commented:
As for the window failure, I actually think that's real.  And I'm wondering if its because of the \\ vs / difference.  @ahmetalpbalkan should the .dockerfile (on windows) have \\ or /'s ?

buddhamagnet commented:
@duglin yep that code is good - I have just moved the variable assignment for matched below that initial check because it's not needed if the first check fails.

buddhamagnet commented:
@duglin yes the Windows failure is for real.

buddhamagnet commented:
@duglin although I'm not sure why TestBuildDockerignore would pass and TestBuildDockerignoreExceptions would pass.

ahmetalpbalkan commented:
Windows dockerfiles still use `/` for paths. We convert them to `\` magically. Yep the failure seems real.

buddhamagnet commented:
@duglin also wonder whether we should also update the docs:

```txt
articles/dockerfile_best-practices.md
reference/builder.md
```

buddhamagnet commented:
@ahmetalpbalkan it's excluded in the .dockerignore so should not exist. Odd that the subsequent test passes and that the test that passes in janky fails on the windows run.

buddhamagnet commented:
@ahmetalpbalkan I just ran the process manually in a container using the exact same Dockerfile as used in the integration test and the same .dockerignore and all works perfectly. In fact, some of the files in the test are redundant so have removed them as the first test excludes the whole folder and the subsequent test works with exclusion exceptions in that folder. No idea why it fails on windows and I don't have an easy way to test that here.

duglin commented:
we may need to use PathSeparator instead of "/" in fileutils - idk

buddhamagnet commented:
@duglin not sure - surely the subsequent test would also fail in that case?

ahmetalpbalkan commented:
I think test is doing the right thing and it just revealed a bug. @buddhamagnet maybe that code path wasnt being tested before?

buddhamagnet commented:
@duglin @ahmetalpbalkan not so sure all is green now.

ahmetalpbalkan commented:
lol, i wonder how that did happen

buddhamagnet commented:
@duglin @ahmetalpbalkan the windows tests seem a little flaky so where do we go from here?

ahmetalpbalkan commented:
it's very interesting I can't see any race conditions that might cause flakiness. let's proceed and if it happens again we'll take a look at the test.

buddhamagnet commented:
well, the PR is clean, all unit and integration tests pass and all appropriate suggested amendments have been made so is there anything left to do? @duglin @ahmetalpbalkan.

ahmetalpbalkan commented:
I haven't really reviewed the PR, please let me know if a windows-related issue arises. :)

camilonova commented:
Hope to see this merged finally.

buddhamagnet commented:
@camiloniva hopefully real soon!

buddhamagnet commented:
Me too @camilonova!

buddhamagnet commented:
@duglin so what happens now it all seems to have gone a bit quiet.

thaJeztah commented:
@buddhamagnet (depending on which timezone you're in), it's weekend, "container camp" was this weekend, and some maintainers are still recovering from the 1.6 release :smile:, so I think, maintainers will be able to review next couple of days. There's lots of time before the 1.7 release, so don't worry

buddhamagnet commented:
thanks @thaJeztah!

duglin commented on pkg/archive/archive.go:
```diff
@@ -449,9 +449,6 @@ func TarWithOptions(srcPath string, options *TarOptions) (io.ReadCloser, error)
 				}
 
 				if skip {
-					if f.IsDir() {
```
Been thinking about the performance impact of this PR and I'm a little bit worried about the removal of this SkipDir stuff.  For very large build contexts this could have a serious impact on our upload speed - especially when you consider the fact that we do certain cleanup type of actions over and over in our Matches() func.  For example, we call filepath.Clean() on each pattern and we do it for all patterns and repeat that for each file in the build context.

What I'd like to propose is this....
* before this for-loop (before line 422) loop over options.ExcludePatterns. For each, call filepath.Clean() and save the results in a new slice - so we don't mess with the original.
* also save the split(...,"/") version of the pattern so we don't need to do it in Match()
* while looping, also detect any ! usage and make a note of it - actually, note if we're *not* using it at all.
* error any !'s all by themselves during this loop instead of in Match()
* if there are no !'s in any patterns then keep/do the `return filepath.SkipDir` on line 453.
* move Match() from fileutils (assuming its not used by anyone else) into archive - mainly because the optimizations I'm talking about here make this func less generic and will make some assumptions that a 'generic' pkg shouldn't make.  If Match() shrinks enough, perhaps even just move it into this func at line 444 instead - but worst case make it a sibling to this one.

thoughts?

buddhamagnet commented:
@dugin `fileutils.Matches` is used in:

```txt
api/client/build.go
builder/evaluator.go
utils/utils.go
```

...so not sure moving in wholesale into archive.go is the right move?

buddhamagnet commented:
@duglin I removed the SkipDir on your suggestion, as I said I'm new to this project and didn't have any knowledge on the possible performance impact. Will try and keep moving this forward but time is limited this week. Maybe a good first step is to concur where the function should sit, move it there and adjust all consuming code, confirm all tests still pass and then focus on the optimizations?

buddhamagnet commented:
@duglin or we give archive its own Matches function for now, and once it's optimised, either collapse it into the main code if small enough or finalise it and determine whether the code in the other sections of the codebase can use it or retain the one in fileutils?

buddhamagnet commented:
@duglin going to do this carefully step by step and push once each step is complete and all tests are passing locally.

buddhamagnet commented:
OK. Copied fileutils.Matches into archive.go and updated archive.go's imports. So both packages have a copy of that code now. Also duplicated the tests in acrhive_test.go. All tests still pass.

buddhamagnet commented:
@duglin next step, iterate over options.ExcludePatterns and populate a new slice of filepath cleaned pattersns, and omit the cleaning in this version of Matches.

buddhamagnet commented:
@duglin so summary so far:

In TarWithOptions we now:

1. Loop over exclusion patterns and store cleaned version in a new slice.
2. Indicate whether we are dealing with any exception rules and if not, re-invoke SkipDir later.
3. Return an error if we find a single !.

So I removed the check for the single exclamation mark in the unit tests in archive_test.go, which means that path is not covered now. The ```tarUntar``` function in archive_test.go throws a ```t.Fatalf``` on error so we need to think about testing that.

buddhamagnet commented:
@duglin have also moved the Exclusion and Empty functions back into fileutils as they are generic and made them exportable.

duglin commented on pkg/archive/archive.go:
```diff
@@ -382,6 +382,56 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
+// and isn't excluded by any of the subsequent patterns.
+func Matches(relFilePath string, patterns []string) (bool, error) {
+
+	if filepath.Clean(relFilePath) == "." {
```
Because this is here, and not in the inner-loop after the removal of the "!", we may need to check for "!." here.

duglin commented on pkg/archive/archive.go:
```diff
@@ -391,6 +441,24 @@ func Tar(path string, compression Compression) (io.ReadCloser, error) {
+		if fileutils.Exclusion(pattern) {
+			if len(pattern) == 1 {
+				logrus.Errorf("Illegal exclusion pattern: %s", pattern)
+				return nil, errors.New("Illegal exclusion pattern: !")
```
Since you remove the unit test for this, go ahead and add an integration-cli test for it instead.

duglin commented on pkg/archive/archive.go:
```diff
@@ -382,6 +382,56 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 }
 
+// Matches returns true if relFilePath matches any of the patterns
+// and isn't excluded by any of the subsequent patterns.
```
Let's add something like:
```
// This func is basically the same as fileutils.Match() but optimized for archive.go.
// This will assume that the inputs have been preprocessed and therefore we 
// don't need to do as much error checking/clean-up. This was done to avoid
// repeating these steps on each file being checked during the archive process.
// The more generic fileutils.Match() can't make these assumptions.
```
so that people know why we duplicated it.
Perhaps even add a comment to the fileutils's one pointing to here so that if
people change one they know to change the other.

buddhamagnet replied:
@duglin I wasn't going to add docs until I knew whether the function could be used in both contexts (hence porting Exclusion and Empty back into fileutils as they are generic).

duglin replied:
ok. Another option is to move the archive Matches into fileutils.go, but call it "OptimizedMatches" or something, then have Matches() do all of the clean-up stuff before calling OptimizedMatches().  But archive calls OptimizedMatches. Then we don't dup code.  just a thought.

buddhamagnet replied:
I'd rather do that once we are happy and have the stable code in place, and all tests passing, with no impact on performance, maybe as a follow up PR?

duglin commented on pkg/archive/archive.go:
```diff
@@ -382,6 +382,56 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
+
+	for _, pattern := range patterns {
+		negative := false
+		if fileutils.Empty(pattern) {
```
Let's remove empty patterns from your `cleanedPattern` slice - then we can remove this if-statement because we know the list won't contain them.

duglin replied:
And remove "!" too.

buddhamagnet replied:
@duglin removing empty patterns, but I can't do any more on this today, sorry.

duglin replied:
If you wouldn't mind the joint contributions, if you give me access I can make code changes too while you're sleeping  :-)   you're choice.

buddhamagnet replied:
I'm more than happy man.

buddhamagnet commented:
@duglin stripping empty patterns from cleanedPatterns now.

buddhamagnet commented:
So @duglin I am happy to refactor further when the dust has settled. Aside from adding an further integration test, what else would we need to definitely do now? I'm about to clock off for the day.

buddhamagnet commented:
@duglin added you, feel free. No ego here.

buddhamagnet commented:
@duglin let me make one more change before I cease.

buddhamagnet commented:
@duglin OK I'm done - I extracted that cleanup logic into a function called cleanPatterns and moved it back into fileutils which should give us a good jumping off point for reworking if necessary. Will add some unit tests for it also.

buddhamagnet commented:
@duglin over and out, I won't add unit tests for CleanPatterns yet in case it gets absorbed or refactored out.

buddhamagnet commented:
@duglin just rebased against master and force pushed to keep things clean.

duglin commented:
@buddhamagnet LOL  so much for "over and out" :-)

buddhamagnet commented:
@duglin you know how it gets under your skin man.

buddhamagnet commented:
@duglin although if we are using a rebase workflow on a shared branch we'll need to be careful I guess. What do you propose? Person makes a change, ensures tests pass, commits and force pushes? Other person deletes their feature branch and checks it out fresh?

duglin commented:
I wasn't going to get to it until later, probably after you go to bed so I don't think we'll need to worry about it.

buddhamagnet commented:
@duglin wilco will check in the morning.

vbatts commented on pkg/fileutils/fileutils.go:
```diff
@@ -1,33 +1,105 @@
 )
 
+func Exclusion(pattern string) bool {
+	return pattern[0] == '!'
```
not sure that it will hit here, but the callers of this look to have used `filepath.Clean()` which does nothing for leading (or trailing) whitespaces. So `Exclusion("  !foo  ")` would return `false`

buddhamagnet replied:
Thanks I've trimmed the patterns in CleanPatterns now before they get passed to Exclude.

duglin commented:
I push a branch called "dug" with some edits to your repo - go ahead and take a look.
I'm seeing a weird error in the test but I'm not sure if its really related to these changes, but probably.
Will have to look more tomorrow.

buddhamagnet commented:
@duglin looks good, odd all tests pass for me locally. Have intregrated the work into this branch. I assume I can remove the debugging statements?

buddhamagnet commented:
@duglin so all passing with that work integrated. Will remove the debugging output.

buddhamagnet commented:
Just added some more unit tests for CleanPatterns as it's staying and removed the debugging output.

buddhamagnet commented:
@duglin and a unit test for the directory splitting behaviour.

buddhamagnet commented:
@duglin it could be argue that the unit tests for Matches now cover everything but CleanPatterns can be tested in isolation so I'm testing it.

buddhamagnet commented:
@duglin have also run through as many scenarios manually in my own container with nested folders, leading and trailing whitespace in exception rules, single exclamation points and empty patterns and all checks out.

buddhamagnet commented:
@duglin so 20 new tests, all manual tests check out, an optimised matcher in fileutils and everything documented. Think we're good to go?

duglin commented:
I love all of the new tests!
I think it looks good.
@vbatts can you look it over to make sure you're comfortable with the new flow.
ping @tiborvass

duglin commented:
@buddhamagnet btw - thanks for all of the effort you put into this - I know it ended up being more than you expected, but I like where we ended up.

buddhamagnet commented:
@duglin it has been an utter pleasure my friend.

vbatts commented:
the archive changes LGTM

calavera commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -3695,6 +3704,56 @@ func TestBuildDockerignore(t *testing.T) {
+!dir/e*
+!dir/dir/foo`,
+	})
+	defer ctx.Close()
```
this `defer` should go after the `if err != nil`

buddhamagnet replied:
Done - could you explain the rationale for me please?

calavera replied:
there might not be object `ctx` when there is an error, causing `ctx.Close()` to panic due to a nil reference.

calavera commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -3675,15 +3675,24 @@ func TestBuildDockerignore(t *testing.T) {
+*.md
+dir`,
 	})
 	defer ctx.Close()
```
do you mind to move this `defer` after the `if err  != nil` in this PR?

buddhamagnet replied:
Also done @calavera.

calavera replied:
cool, thank you! :fireworks:

buddhamagnet commented:
@tiborvass any comments?

duglin commented:
@buddhamagnet sorry- rebase needed

buddhamagnet commented:
@duglin rebased and force pushed.

buddhamagnet commented:
@duglin hmm that didn't go well, I'll sort it when I can.

duglin commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -1244,8 +1256,8 @@ To see how the `docker:apache` image was added to a container's base image:
     $ docker history docker:scm
     IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
     2ac9d1098bf1        3 months ago        /bin/bash                                       241.4 MB            Added Apache to Fedora base image
-    88b42ffd1f7c        5 months ago        /bin/sh -c #(nop) ADD file:1fd8d7f9f6557cafc7   373.7 MB            
```
Let's undo these edits and the ones below.

buddhamagnet commented:
@duglin unless you have the latest copy of this branch right now in which case feel free to force push it as I am out drinking beer later. And I will handle the rebase tomorrow.

duglin commented:
I don't think I do so we'll just wait.
maintainers: the current code in here is old - please do not review it yet.  vbatts' review was on the right code base so his LGTM was accurate.

buddhamagnet commented:
@duglin OK rebased up to date codebase, brought new integration test in line with new function format and force pushed.

duglin commented:
LGTM
ping @tiborvass @LK4D4

buddhamagnet commented:
@duglin do you know if there's any way to get this moving along?

duglin commented:
just waiting on one more review - will poke on Monday

buddhamagnet commented:
@duglin thanks man!

buddhamagnet commented:
@duglin rebased and pushed.

calavera commented:
Code and design LGTM. Moving to docs review :tada:

moxiegirl commented on docs/sources/reference/builder.md:
```diff
@@ -46,6 +46,8 @@ the files needed for building that Dockerfile to that directory. To further spee
 build, you can exclude files and directories by adding a `.dockerignore` file to the same
 directory.
 
+The `.dockerignore` file also supports exclusion patterns (as per `.gitignore` files).
```
Combine this with the paragraph above. Also, are the patterns the same or similar? If they are the same we should say so. If they are similar, we should document the differences or point to where they exist. I'm assuming they are the same. So,

---> https://gist.github.com/moxiegirl/2a8f455f0e7836377457

In most cases, it's best to put each Dockerfile in an empty directory. Then,
only add the files needed for building the Dockerfile to the directory. To
increase the build's performance, you can exclude files and directories by
adding a `.dockerignore` file to the directory. This file supports the same exclusion patterns as `.gitignore` files.

moxiegirl commented on docs/sources/articles/dockerfile_best-practices.md:
```diff
@@ -40,6 +40,8 @@ context and final image. For example, unless`.git` is needed by your build
 process or scripts, you should add it to `.dockerignore`, which can save many
 megabytes worth of upload time.
 
+The `.dockerignore` file also supports exclusion patterns (as per `.gitignore` files).
```
Add to the paragraph above.  Also, reduce complexity in the original paragraph (not your fault).

--> https://gist.github.com/moxiegirl/88f830e38843be32a1df

For faster uploading and efficiency during `docker build`, you should use
a `.dockerignore` file. This file excludes files or directories from the build context and final image. For example, unless`.git` is needed by your build
process or scripts, you should add it to `.dockerignore`. This can save many
megabytes worth of upload time. The `.dockerignore` file also supports the same exclusion patterns as a `.gitignore` file.

buddhamagnet commented:
@moxiegirl amended and pushed.

moxiegirl commented:
@buddhamagnet You are super quick.  I noticed a few problems in the way we document this file. Unfortunately, after my initial comments. My apologies. I'll sort this out and post a Gist with the changes -- if you are willing to make them.

buddhamagnet commented:
@moxiegirl go for it.

moxiegirl commented:
Hi @buddhamagnet -- I ended up creating a Gist containing a patch here: https://gist.github.com/moxiegirl/d0ec84384df2763b82fd

Just download it and apply it to your branch.

buddhamagnet commented:
@moxiegirl output from dummy run:
```txt
git apply --stat moxie.patch
 docs/sources/articles/dockerfile_best-practices.md |   12 +-
 docs/sources/reference/builder.md                  |   97 +++++++++++++-------
 docs/sources/reference/commandline/cli.md          |   68 +++++---------
```

output from dummy run:

```txt
git apply --check moxie.patch
error: patch failed: docs/sources/articles/dockerfile_best-practices.md:32
error: docs/sources/articles/dockerfile_best-practices.md: patch does not apply
error: patch failed: docs/sources/reference/builder.md:41
error: docs/sources/reference/builder.md: patch does not apply
error: patch failed: docs/sources/reference/commandline/cli.md:674
error: docs/sources/reference/commandline/cli.md: patch does not apply
```

Looks like that patch won't apply cleanly?

buddhamagnet commented:
@moxiegirl also, it looks like a lot of these changes pertain to the documentation of the dockerignore file, rather than changes introduced by this feature. Could this not be added in a separate pull request (just asking).

camilonova commented:
Man, is so hard to make this happen, should be easier. @buddhamagnet thank you for keeping pushing forward.

buddhamagnet commented:
@camilonova @moxiegirl it's not hard at all, I'm happy ;)

moxiegirl commented:
@buddhamagnet Normally, I wouldn't ask for this many changes which is why I checked with you yesterday to see if you are willing. Since all your documentation changes were around the `.dockerignore` file, I think it is appropriate to clean up the whole slew with the same PR.

buddhamagnet commented:
@moxiegirl no worries at all. Easier for me to give you remote access and you issue a PR?

buddhamagnet commented:
@moxiegirl may well be because I rebased against master and force pushed just before I got the patch. We can either try a new patch or your can issue a PR against my branch, whichever is easiest and spreads more rainbows and unicorns.

moxiegirl commented:
Sure thing. Here you go: https://github.com/buddhamagnet/docker/pull/2 @buddhamagnet

buddhamagnet commented:
@moxiegirl yeah some conflicts there, resolved them, merged and rebased my own branch to combine ir all into the original commit etc. Hope it's good.

buddhamagnet commented:
@moxiegirl all incorporated, all tests green, one single shining commit.

calavera commented:
:tada:

thaJeztah commented:
\o/ awesome! Thanks for the contribution!

buddhamagnet commented:
HOLY COW PARTY TIME!!!

buddhamagnet commented:
@duglin you needed some credit on that commit.

duglin commented:
@buddhamagnet just glad to help move it along
